### ISSUE TITLE
False positive when assigning to `Final` instance attribute in `__init__`

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1409]
From the conformance suite (and 44 instances across the ecosystem once type-of-self lands):

```py
from typing import Final, Self

class ClassA:
    ID4: Final[int]  # OK because initialized in __init__

    def __init__(self: Self):
        self.ID4 = 1  #     Cannot assign to final attribute `ID4` on type `Self@__init__` (invalid-assignment)
```
https://play.ty.dev/38bf9854-4d2e-495c-b96b-29553d212819

**Comment by saada** [comment]
I've submitted a PR to fix this issue: https://github.com/astral-sh/ruff/pull/21158

The fix allows Final instance attributes to be initialized in `__init__` methods as per PEP 591.
